goshopify.go

```diff
@@ -308,17 +297,35 @@ func (c *Client) Do(req *http.Request, v interface{}) error {
 
 // doGetHeaders executes a request, decoding the response into `v` and also returns any response headers.
 func (c *Client) doGetHeaders(req *http.Request, v interface{}) (http.Header, error) {
-	resp, err := c.Client.Do(req)
-	if err != nil {
-		return nil, err
-	}
-	defer resp.Body.Close()
+	var resp *http.Response
+	var err error
+	for {
+		resp, err = c.Client.Do(req)
+		if err != nil {
+			return nil, err
+		}
 
-	err = CheckResponseError(resp)
-	if err != nil {
-		return nil, err
+		err = CheckResponseError(resp)
+		if err != nil {
+			if retry, isRetryErr := err.(RateLimitError); isRetryErr {
+				// back off and retry
+				if c.retries <= 0 {
+					return nil, err
+				}
+				wait := time.Duration(retry.RetryAfter) * time.Second
+				time.Sleep(wait)
+				c.retries--
+				continue
+			}
+			return nil, err
+		}
+		break
 	}
 
+	defer func() {
+		err = resp.Body.Close()
```

Comment:
why not just `defer resp.Body.Close()` ?

Reply:
it returns an error, just want to capture it. im open to leaving it if you dont want to mess with it.

Reply:
But you're not doing anything with that error (you assign it to `err` but there's evaluated after the function is done executing and `err` is returned)

Reply:
does it not return that err instead? got the idea from a blog tbh: https://play.golang.org/p/PZGn5-1TXv4

Reply:
See this simplified playground example: https://play.golang.org/p/4wJwo5UO5Rz

^^^ always returns nil for me

Reply:
ya so my understanding was wrong, i'd need to do the named param. i can update to the named param or remove the func() entirely.

Reply:
👍  - I'd remove the func and simplify things.

Reply:
https://play.golang.org/p/UyEzYmPItcD

just always find it weird that everyone is ok with eating errors from defer'ed closes and wanted to try this method. /shrug

Reply:
updated this code to do the local var, removed the func() and added an extra close in the retry logic to be safe. give me a high five on this and ill update the tests.